### ISSUE TITLE
Faster $driver$get_hash() for the RDS driver

### ISSUE THREAD

**Issue by wlandau** [richfitz/storr#96]
I am trying to speed up `drake` (ref: https://github.com/ropensci/drake/issues/647, https://github.com/ropensci/drake/issues/655). Note where `self_hash()` appears in the middle of [this flame graph](https://user-images.githubusercontent.com/1580860/50739738-4de1b800-11b2-11e9-8695-b1cac231f4b5.png). Candidate alternatives to `readLines()`: https://gist.github.com/hadley/6353939.

Would you be open to a PR that dropped down into C?

**Comment by wlandau** [comment]
Hmm... the speed differences from https://gist.github.com/hadley/6353939 do seem really nice at first glance.

``` r
path <- file.path(R.home("doc"), "COPYING")
size <- file.info(path)$size
microbenchmark::microbenchmark(
  old = readLines(path),
  new = readChar(path, size, useBytes = TRUE)
)
#> Unit: microseconds
#>  expr     min       lq      mean   median       uq     max neval cld
#>   old 224.137 226.2785 243.32306 232.1250 255.6135 302.260   100   b
#>   new  61.374  64.3230  70.43012  67.8185  72.3140 124.196   100  a
```

However, at least for `readChar()`, the difference is less impressive for the small key files that `storr` uses. Part of the speedup below could just be from precomputing the hash length.

``` r
writeLines(digest::digest(NA, "md5"), "key")
path <- "key"
microbenchmark::microbenchmark(
  old = readLines(path),
  new = readChar(path, nchars = 32, useBytes = FALSE)
)
#> Unit: microseconds
#>  expr    min      lq     mean  median      uq     max neval cld
#>   old 32.795 35.1310 37.86379 36.4545 37.7230  59.820   100   b
#>   new 24.267 25.8445 28.56502 26.5890 27.6045 171.191   100  a
```

I will still open a PR in case you think it is worth the change.

**Comment by wlandau** [comment]
One thing I wondered: is it possible to go any faster if we drop even farther down to pure C? Overly optimistic C function:

```c
// read.c
void readkey_c() {
  char buf[32];
  FILE *fp;
  fp = fopen("key", "rb");
  fread(buf, 32, 1, fp);
  //  printf("%s", buf);
  fclose(fp);
  return;
}
```

Benchmarks:

``` r
writeLines(digest::digest(NA, "md5"), "key")
path <- "key"
Rcpp::sourceCpp("~/read-file.cpp")
dyn.load("~/read.so") # R CMD SHLIB read.c -i read.so
microbenchmark::microbenchmark(
  old = readLines(path),
  new = readChar(path, nchars = 32, useBytes = FALSE),
  cpp1 = read_file_cpp1(path),
  cpp2 = read_file_cpp1(path),
  c = .C("readkey_c")
)
#> Unit: microseconds
#>  expr    min     lq      mean  median      uq       max neval cld
#>   old 35.464 38.750  43.38168 40.7825 43.3590   129.943   100   a
#>   new 25.399 29.286  31.87871 30.7565 31.9520    56.339   100   a
#>  cpp1  7.128  8.148 161.75054 11.0595 12.4700 15076.276   100   a
#>  cpp2  7.196  8.290  11.45663 10.9160 12.4450    81.264   100   a
#>     c  6.968  9.398  10.89305 10.4880 11.1485    32.054   100   a
```

<sup> Created on 2019-01-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1) </sup>

The performance bump from dropping down to C is negligible. I think the best we can do is this:

```c
#include <fstream>
#include <sstream>
#include <string>
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
CharacterVector read_file_cpp1(std::string path) {
  std::ifstream t(path.c_str());
  std::stringstream ss;
  ss << t.rdbuf();
  return ss.str();
}
```

**Comment by richfitz** [comment]
Fixed in #98
